lib/runtime-core/src/trampoline_x64.rs

```diff
@@ -0,0 +1,245 @@
+    static ref GET_CONTEXT: extern "C" fn () -> *const CallContext = {
+        static CODE: &'static [u8] = &[
+            0x48, 0x0f, 0x7e, 0xc0, // movq %mm0, %rax
+            0xc3, // retq
+        ];
+        let mut mem = CodeMemory::new(4096);
+        mem[..CODE.len()].copy_from_slice(CODE);
+        mem.make_executable();
+        let ptr = mem.as_ptr();
+        mem::forget(mem);
+        unsafe {
+            mem::transmute(ptr)
+        }
+    };
+}
+
+/// An opaque type for pointers to a callable memory location.
+pub enum CallTarget {}
+
+/// An opaque type for context pointers.
+pub enum CallContext {}
+
+/// An opaque type for generated trampolines' call entries.
+pub enum Trampoline {}
+
+/// Trampoline Buffer Builder.
+pub struct TrampolineBufferBuilder {
+    code: Vec<u8>,
+    offsets: Vec<usize>,
+}
+
+/// Trampoline Buffer.
+pub struct TrampolineBuffer {
+    code: CodeMemory,
+    offsets: Vec<usize>,
+}
+
+fn value_to_bytes<T: Copy>(ptr: &T) -> &[u8] {
+    unsafe { slice::from_raw_parts(ptr as *const T as *const u8, mem::size_of::<T>()) }
+}
+
+/// Calls `GET_CONTEXT` and returns the current context.
+pub fn get_context() -> *const CallContext {
+    GET_CONTEXT()
+}
+
+impl TrampolineBufferBuilder {
+    pub fn new() -> TrampolineBufferBuilder {
+        TrampolineBufferBuilder {
+            code: vec![],
+            offsets: vec![],
+        }
+    }
+
+    /// Adds a context trampoline.
+    ///
+    /// This generates a transparent trampoline function that forwards any call to `target` with
+    /// unmodified params/returns. When called from the trampoline, `target` will have access to
+    /// the `context` specified here through `get_context()`.
+    ///
+    /// Note that since `rax` is overwritten internally, variadic functions are not supported as `target`.
+    pub fn add_context_trampoline(
+        &mut self,
+        target: *const CallTarget,
+        context: *const CallContext,
+    ) -> usize {
+        let idx = self.offsets.len();
+        self.offsets.push(self.code.len());
+        self.code.extend_from_slice(&[
+            0x48, 0xb8, // movabsq ?, %rax
+        ]);
+        self.code.extend_from_slice(value_to_bytes(&context));
+        self.code.extend_from_slice(&[
+            0x48, 0x0f, 0x6e, 0xc0, // movq %rax, %mm0
+        ]);
+        self.code.extend_from_slice(&[
+            0x48, 0xb8, // movabsq ?, %rax
+        ]);
+        self.code.extend_from_slice(value_to_bytes(&target));
+        self.code.extend_from_slice(&[
+            0xff, 0xe0, // jmpq *%rax
+        ]);
+        idx
+    }
+
+    /// Adds a callinfo trampoline.
+    ///
+    /// This generates a trampoline function that collects `num_params` parameters into an array
+    /// and passes the array into `target` as the second argument when called. The first argument
+    /// of `target` is the `context` specified here.
+    ///
+    /// Note that non-integer parameters/variadic functions are not supported.
```

Comment:
Does it support floats (I think it does not matter since the bits can be interpreted as a float, but I prefer to ask)?

Reply:
Floats are passed in different registers and it would be very complex to handle this case. I think it's better to just support integer parameters.

Reply:
So trampoline doesn't work with floats. It's important to know!